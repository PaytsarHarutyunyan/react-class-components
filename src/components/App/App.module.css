* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}

.container {
    width: 90%;
    margin: auto;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
}

.content {
    display: flex;
    justify-content: space-between;
    flex-direction: column;
    border-radius: 10px;
    width: 100%;
    height: 600px;
    background-color: #dcdcdc;
}

.key {
    display: inline-block;
}