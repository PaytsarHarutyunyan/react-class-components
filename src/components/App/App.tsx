import React, { useState, useEffect } from 'react';
import styles from './App.module.css';
import { getData } from '../../api/index';
import { BASE_URL } from '@/constants/commonConstants';
import useLocalStorage from '../../hooks/useLocalStorage';
import Routes from '../../routes/Routes';
import { useSelectedPage } from '../../hooks/useSelectedPage';
import { Result, Tab } from '@/types';

const App: React.FC = () => {
    const [loading, setLoading] = useState<boolean>(false);
    const [result, setResult] = useState<Result>({ count: 0, data: [] });
    const [perPageCount] = useState<number>(10);
    const [selectedTab, setSelectedTab] = useState<Tab>({ name: 'People', url: '/people' });
    const [tabs] = useState<Tab[]>([
        { name: 'People', url: '/people' },
        { name: 'Planets', url: '/planets' },
        { name: 'Species', url: '/species' },
        { name: 'Starships', url: '/starships' },
        { name: 'Vehicles', url: '/vehicles' },
    ]);
    const [itemDetails] = useState<{ [key: string]: string } | null>(null);
    const [searchTerm, setSearchTerm] = useLocalStorage<string>('searchTerm', '');

    const { getSelectedPage } = useSelectedPage();
    const selectedPage = getSelectedPage();

    const fetchData = async (url: string, options?: { searchTerm?: string; page?: number }) => {
        setLoading(true);
        const { searchTerm, page } = options || {};

        const queryParams: Record<string, string> = {};
        if (searchTerm) queryParams['search'] = searchTerm;
        if (page) queryParams['page'] = String(page);

        const searchParams = new URLSearchParams(queryParams);
        const result = await getData(`${BASE_URL}${url}?${searchParams.toString()}`);
        setLoading(false);
        setResult({ data: result.results, count: result.count });
    };

    useEffect(() => {
        fetchData(selectedTab.url, { page: selectedPage });
    }, [selectedTab.url, selectedPage]);

    const changeTab = async (tabName: string) => {
        const tabToSet = tabs.find((tab) => tab.name === tabName) || selectedTab;
        await fetchData(tabToSet.url, { searchTerm });
        setSelectedTab(tabToSet);
    };

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const onClickItem = async () => {
        // const result = await fetchData(`${selectedTab.url}/${id}`);
        // setItemDetails(result);
    };

    const onClickPagination = async (page: number) => {
        await fetchData(selectedTab.url, { searchTerm, page });
    };

    const handleSearch = async (searchTerm: string) => {
        setSearchTerm(searchTerm);
        await fetchData(selectedTab.url, { searchTerm });
    };

    const drawSingleResult = () => {
        if (!itemDetails) return null;
        return Object.keys(itemDetails).map((key) => (
            <span key={key}>
                <h4 className={styles.key}>{key}:</h4>
                <span>{itemDetails[key]}</span>
            </span>
        ));
    };

    return (
        <div className={styles.hero}>
            <div className={styles.container}>
                <div className={styles.content}>
                    <Routes
                        tabs={tabs}
                        selectedTab={selectedTab}
                        result={result}
                        loading={loading}
                        singleResult={itemDetails}
                        perPageCount={perPageCount}
                        searchTerm={searchTerm}
                        handleSearch={handleSearch}
                        headerBtnAction={changeTab}
                        onClickItem={onClickItem}
                        onClickPagination={onClickPagination}
                        drawSingleResult={drawSingleResult}
                    />
                </div>
            </div>
        </div>
    );
};

export default App;
