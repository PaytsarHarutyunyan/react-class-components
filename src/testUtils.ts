import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithRouter = (ui: React.ReactElement) => {
    return render(ui, { wrapper: MemoryRouter });
};
