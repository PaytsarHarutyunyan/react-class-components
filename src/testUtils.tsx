import React, { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from './store';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from './context/ThemeProvider';

export const renderWithProvidersAndRouter = (ui: ReactElement, { route = '/' } = {}) => {
    return render(
        <Provider store={store}>
            <ThemeProvider>
                <MemoryRouter initialEntries={[route]}>{ui}</MemoryRouter>
            </ThemeProvider>
        </Provider>,
    );
};
