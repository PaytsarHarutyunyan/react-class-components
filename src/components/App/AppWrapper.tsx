import App from './App';
import { BrowserRouter as Router } from 'react-router-dom';
import ErrorBoundary from '../ErrorBoundary/ErrorBoundary';

const AppWrapper = () => {
    return (
        <Router>
            <ErrorBoundary>
                <App />
            </ErrorBoundary>
        </Router>
    );
};

export default AppWrapper;
