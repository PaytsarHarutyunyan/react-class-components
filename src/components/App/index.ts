import App from '../../pages';
export default App;
