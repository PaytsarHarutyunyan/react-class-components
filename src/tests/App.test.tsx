import React from 'react';
import { screen, waitFor } from '@testing-library/react';
import App from '../components/App/App';
import fetchMock from 'jest-fetch-mock';
import { renderWithProvidersAndRouter } from '@/testUtils';
import peopleMock from './mocks/peopleMock.json';

// Initialize fetchMock
fetchMock.enableMocks();

const mockTabs = [
    { name: 'People', url: '/people' },
    { name: 'Planets', url: '/planets' },
    { name: 'Species', url: '/species' },
    { name: 'Starships', url: '/starships' },
    { name: 'Vehicles', url: '/vehicles' },
];

describe('App Component', () => {
    beforeEach(() => {
        fetchMock.resetMocks(); // Reset mocks before each test
    });

    test('should render correctly', () => {
        fetchMock.mockResponse(
            JSON.stringify({ results: peopleMock, count: 10, prev: null, next: null }),
        );
        const { container } = renderWithProvidersAndRouter(<App />);

        expect(container).toMatchSnapshot();
    });

    test('renders header buttons', async () => {
        fetchMock.mockResponse(
            JSON.stringify({ results: peopleMock, count: 10, prev: null, next: null }),
        );

        renderWithProvidersAndRouter(<App />);
        mockTabs.forEach(async (tab) => {
            await waitFor(() => expect(screen.getByText(tab.name)).toBeInTheDocument());
        });
    });
});
