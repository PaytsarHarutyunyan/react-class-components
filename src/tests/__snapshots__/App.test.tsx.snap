// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render correctly 1`] = `
<div>
  <div
    class="hero"
  >
    <div
      class="container"
    >
      <div
        class="content"
      >
        <div
          class="searchContainer"
        >
          <input
            placeholder="Search..."
            type="text"
            value=""
          />
          <button>
            Search
          </button>
        </div>
        <div
          class="content"
        >
          <div
            class="container"
          >
            <button
              class="item itemSelected"
            >
              People
            </button>
            <button
              class="item"
            >
              Planets
            </button>
            <button
              class="item"
            >
              Species
            </button>
            <button
              class="item"
            >
              Starships
            </button>
            <button
              class="item"
            >
              Vehicles
            </button>
          </div>
          <span>
            Loading...
          </span>
          <div
            class="container"
          >
            <div
              class="itemContainer"
            >
              <button
                class="item"
              >
                <a
                  href="/?page=1"
                  style="display: block; color: rgb(255, 255, 255); text-decoration: none;"
                >
                  &lt;&lt;
                </a>
              </button>
              <button
                class="item"
              >
                <a
                  href="/?page=1"
                  style="display: block; color: rgb(255, 255, 255); text-decoration: none;"
                >
                  &lt;
                </a>
              </button>
              No data available
              <button
                class="item"
              >
                <a
                  href="/?page=2"
                  style="display: block; color: rgb(255, 255, 255); text-decoration: none;"
                >
                  &gt;
                </a>
              </button>
              <button
                class="item"
              >
                <a
                  href="/?page=0"
                  style="display: block; color: rgb(255, 255, 255); text-decoration: none;"
                >
                  &gt;&gt;
                </a>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
